File: examples/python/RVC4/GPUStereo/gpu_stereo.py
#!/usr/bin/env python3
"""Minimal GPUStereo example — disparity from a stereo camera pair (RVC4 only)."""

import cv2
import depthai as dai
import numpy as np

device = dai.Device()

if not device.isGpuAvailable():
    print("Exiting GPUStereo example: GPU not available on this device.")
    raise SystemExit(0)

pipeline = dai.Pipeline(device)
cam_left = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_B)
cam_right = pipeline.create(dai.node.Camera).build(dai.CameraBoardSocket.CAM_C)

gpu = pipeline.create(dai.node.GPUStereo)
cam_left.requestOutput((1280, 800), type=dai.ImgFrame.Type.GRAY8, fps=30).link(gpu.left)
cam_right.requestOutput((1280, 800), type=dai.ImgFrame.Type.GRAY8, fps=30).link(gpu.right)
gpu.setRectification(True)
gpu.setConfidenceThreshold(25)

disp_q = gpu.disparity.createOutputQueue()

with pipeline:
    pipeline.start()
    while pipeline.isRunning():
        frame = disp_q.get()
        disp = frame.getFrame()

        disp8 = (disp.astype(np.float32) * (255.0 / 96.0)).clip(0, 255).astype(np.uint8)
        disp8 = cv2.applyColorMap(disp8, cv2.COLORMAP_JET)

        cv2.imshow("GPUStereo Disparity", disp8)
        if cv2.waitKey(1) == ord("q"):
            break

File: examples/cpp/GPUStereo/gpu_stereo.cpp
/**
 * Minimal GPUStereo example — disparity from a stereo camera pair (RVC4 only).
 */
#include <depthai/depthai.hpp>
#include <iostream>
#include <opencv2/opencv.hpp>

int main() {
    dai::Device device;

    if(!device.isGpuAvailable()) {
        std::cout << "Exiting GPUStereo example: GPU not available on this device.\n";
        return 0;
    }

    dai::Pipeline pipeline(device);
    auto camLeft = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_B);
    auto camRight = pipeline.create<dai::node::Camera>()->build(dai::CameraBoardSocket::CAM_C);

    auto gpu = pipeline.create<dai::node::GPUStereo>();
    camLeft->requestOutput({1280, 800}, dai::ImgFrame::Type::GRAY8, 30)->link(gpu->left);
    camRight->requestOutput({1280, 800}, dai::ImgFrame::Type::GRAY8, 30)->link(gpu->right);
    gpu->setRectification(true);
    gpu->setConfidenceThreshold(25);

    auto dispQ = gpu->disparity.createOutputQueue();

    pipeline.start();
    while(pipeline.isRunning()) {
        auto frame = dispQ->get<dai::ImgFrame>();
        cv::Mat disp(frame->getHeight(), frame->getWidth(), CV_16UC1, frame->getData().data());
        cv::Mat dispF;
        disp.convertTo(dispF, CV_32F);
        cv::Mat disp8;
        cv::normalize(dispF, disp8, 0, 255, cv::NORM_MINMAX, CV_8U);
        cv::Mat dispColor;
        cv::applyColorMap(disp8, dispColor, cv::COLORMAP_JET);
        cv::imshow("GPUStereo Disparity", dispColor);
        if(cv::waitKey(1) == 'q') break;
    }

    return 0;
}
